File: apps/chat/libs/chat-storage/src/main/java/com/example/chat/storage/mapper/ChannelMapper.java
package com.example.chat.storage.mapper;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import org.springframework.stereotype.Component;

import com.example.chat.domain.channel.Channel;
import com.example.chat.domain.channel.ChannelId;
import com.example.chat.domain.user.UserId;
import com.example.chat.storage.entity.ChatChannelEntity;
import com.example.chat.storage.entity.ChatChannelMemberEntity;

/**
 * Channel Domain ↔ ChatChannelEntity 변환
 */
@Component
public class ChannelMapper {

        /**
         * Domain → Entity 변환 (멤버 정보 제외)
         */
        public ChatChannelEntity toEntity(Channel channel) {
                return ChatChannelEntity.builder()
                                .id(channel.getId().value())
                                .name(channel.getName())
                                .description(channel.getDescription())
                                .channelType(channel.getType())
                                .ownerId(channel.getOwnerId().value())
                                .active(channel.isActive())
                                .createdAt(channel.getCreatedAt())
                                .updatedAt(channel.getUpdatedAt())
                                .build();
        }

        /**
         * Domain의 멤버 정보를 ChatChannelMemberEntity 리스트로 변환
         */
        public List<ChatChannelMemberEntity> toMemberEntities(Channel channel) {
                return channel.getMemberIds().stream()
                                .map(userId -> ChatChannelMemberEntity.builder()
                                                .channelId(channel.getId().value())
                                                .userId(userId.value())
                                                .joinedAt(java.time.Instant.now())
                                                .build())
                                .collect(Collectors.toList());
        }

        /**
         * Entity → Domain 변환
         */
        public Channel toDomain(ChatChannelEntity entity, Set<String> memberIds) {
                Set<UserId> userIds = memberIds.stream()
                                .map(UserId::of)
                                .collect(Collectors.toSet());

                return Channel.fromStorage(
                                ChannelId.of(entity.getId()),
                                entity.getName(),
                                entity.getDescription(),
                                entity.getChannelType(),
                                UserId.of(entity.getOwnerId()),
                                userIds,
                                entity.isActive(),
                                entity.getCreatedAt(),
                                entity.getUpdatedAt());
        }
}

File: apps/chat/libs/chat-storage/src/main/kotlin/com/example/chat/storage/mapper/ChannelMapper.kt
package com.example.chat.storage.mapper

import com.example.chat.domain.channel.Channel
import com.example.chat.domain.channel.ChannelId
import com.example.chat.domain.user.UserId
import com.example.chat.storage.entity.ChatChannelEntity
import com.example.chat.storage.entity.ChatChannelMemberEntity
import org.springframework.stereotype.Component

/**
 * Channel Domain ↔ ChatChannelEntity 변환
 */
@Component
class ChannelMapper {

    /**
     * Domain → Entity 변환 (멤버 정보 제외)
     */
    fun toEntity(channel: Channel): ChatChannelEntity {
        return ChatChannelEntity(
            id = channel.id.value,
            name = channel.name,
            description = channel.description,
            channelType = channel.type,
            ownerId = channel.ownerId.value,
            active = channel.active,
            createdAt = channel.createdAt,
            updatedAt = channel.updatedAt
        )
    }

    /**
     * Domain의 멤버 정보를 ChatChannelMemberEntity 리스트로 변환
     */
    fun toMemberEntities(channel: Channel): List<ChatChannelMemberEntity> {
        return channel.memberIds.map { userId ->
            ChatChannelMemberEntity(
                channelId = channel.id.value,
                userId = userId.value
            )
        }
    }

    /**
     * Entity → Domain 변환
     */
    fun toDomain(entity: ChatChannelEntity, memberIds: Set<String>): Channel {
        val userIds = memberIds.map { UserId.of(it) }.toSet()

        return Channel.fromStorage(
            id = ChannelId.of(entity.id),
            name = entity.name,
            description = entity.description,
            type = entity.channelType,
            ownerId = UserId.of(entity.ownerId),
            memberIds = userIds,
            active = entity.active,
            createdAt = entity.createdAt,
            updatedAt = entity.updatedAt
        )
    }
}

